### PR TITLE
HTML Result: make it a real plugin [v4]

### PR DESCRIPTION
**Pull request by clebergnu**
This work rewrites a part of the HTML report generation code, and then moves it out of the `avocado.core`, removing knowledge and dependencies about the HTML code.

--

Some of the code on the original PR have already been merged, but this is the PR history:

Changes from v3 (#1385):
 * Use the `optional_plugins` directory
 * Use a "no copy" version of the test result directory dict
 * Changed names of the report model methods and template references to better match the result attributes/methods

Changes from v2 (#1375):
 * Fixed a reference to `results_dir` that was left in the debugging output 

Changes from v1 (#1372):
 * Fixed the module namespace used to look for resources

### CODE REVIEW

**Comment by clebergnu** [comment]
@ldoktor @apahim this needs your attention.

FIY, I've tested up to generating RPMs and installing the `avocado` package only, which fixes this issue:
https://trello.com/c/dq26jqQJ/770-bug-avocado-plugins-html-rpm-still-does-not-behave-as-it-should

IIRC, @ldoktor had already reviewed most of it.

Thanks.

File: avocado/core/html.py [hunk 1]
```diff
@@ -130,20 +130,21 @@ def tests(self):
                    "INTERRUPTED": "danger"}
         test_info = []
         results_dir = self.results_dir(False)
-        for tst in self.result.tests:
-            tst = tst.copy()    # we don't want to override other's results
-            tst["test"] = str(tst["name"])
-            logdir = os.path.join(results_dir, 'test-results', tst['logdir'])
-            tst['logdir'] = os.path.relpath(logdir, self.html_output_dir)
+        for t in self.result.tests:
```
Comment (ldoktor) [inline avocado/core/html.py]: well it'd have been better to keep the `tst` name here, but lets see what's further in the code... We can do pep8 fixes later...

File: avocado/core/html.py [hunk 2]
```diff
@@ -130,20 +130,21 @@ def tests(self):
...
+            formatted = {}
+            formatted['name'] = t['name']
```
Comment (ldoktor) [inline avocado/core/html.py]: it'd be safer to use `str` here, but I'm fine with this if it works fine (I'll check it in functional review later)

File: Makefile [hunk 3]
```diff
@@ -129,7 +129,7 @@ requirements-plugins: requirements
 		do AVOCADO_DIRNAME=$(AVOCADO_DIRNAME) make -C $$MAKEFILE requirements &>/dev/null && echo ">> DONE $$MAKEFILE" || echo ">> SKIP $$MAKEFILE";\
 	done
 
-smokecheck:
+smokecheck: clean develop
 	./scripts/avocado run passtest.py
```
Comment (ldoktor) [inline Makefile]: How about removing the smokecheck completely? It was originally intended for travis check but we decided to use the full `check` instead. Based on the first commit here I guess nobody uses it as it hasn't been working for quite a while...

File: setup.py [hunk 4]
```diff
@@ -111,8 +111,6 @@ def get_long_description():
           url='http://avocado-framework.github.io/',
           use_2to3=True,
           packages=find_packages(exclude=('selftests*',)),
```
Comment (ldoktor) [inline setup.py]: At first I was afraid this would include the `optional_plugins`, but it ignores it...

**Comment by ldoktor** [comment]
Hello @clebergnu, I went through all the commits, there was just one minor style issue (too short variable name), but it's acceptable and we can address that in separate. I tested it extensively, it fixes the RPM issue as well as the multiple opened tabs when `--html` is in use. I hope I haven't missed anything crucial as it's quite a medium-core patch set, but it seems to be working well, so let's get this merged.

File: avocado/core/html.py (hunk 2 above)
Comment (ldoktor) [inline avocado/core/html.py]: Works well so let's keep it...

File: avocado/core/html.py (hunk 1 above)
Comment (ldoktor) [inline avocado/core/html.py]: I sent a PR to fix that...

**Pull request merged**